runtime/memprof.c

```diff
@@ -598,52 +622,88 @@ void caml_memprof_track_young(tag_t tag, uintnat wosize, int from_caml)
-
   if (!from_caml) {
+    unsigned n_samples = 1 +
+      mt_generate_binom(caml_memprof_young_trigger - 1 - Caml_state->young_ptr);
+    CAMLassert(encoded_alloc_lens == NULL);    /* No Comballoc in C! */
     caml_memprof_renew_minor_sample();
 
     callstack = capture_callstack_postponed();
-    if (callstack == 0) return;
+    if (callstack == 0) { CAMLreturn0; }
 
     new_tracked(n_samples, Make_header(wosize, tag, Caml_white),
-                0, 1, Val_hp(Caml_state->young_ptr), callstack);
+                0, 1, 0, Val_hp(Caml_state->young_ptr), callstack);
     caml_memprof_check_action_pending();
-    return;
+    CAMLreturn0;
   }
 
-  /* We need to call the callback for this sampled block. Since the
+  /* We need to call the callbacks for this sampled block. Since each
      callback can potentially allocate, the sampled block will *not*
      be the one pointed to by [caml_memprof_young_trigger]. Instead,
      we remember that we need to sample the next allocated word,
      call the callback and use as a sample the block which will be
      allocated right after the callback. */
 
-  /* Restore the minor heap in a valid state for calling the callback.
+  CAMLassert(Caml_state->young_ptr < caml_memprof_young_trigger &&
+             caml_memprof_young_trigger <= Caml_state->young_ptr + whsize);
+  trigger_hp = (header_t*)caml_memprof_young_trigger;
+  young_hp = (header_t*)Caml_state->young_ptr;
+  alloc_hp = young_hp + whsize;
+
+  /* Restore the minor heap in a valid state for calling the callbacks.
      We should not call the GC before these two instructions. */
   Caml_state->young_ptr += whsize;
   caml_memprof_renew_minor_sample();
-
   caml_memprof_suspended = 1;
-  callstack = capture_callstack();
-  t_idx = new_tracked(n_samples, Make_header(wosize, tag, Caml_white),
-                      0, 1, Placeholder_value, callstack);
-  if (t_idx == Invalid_index)
-    res = Val_unit;
-  else
-    res = handle_entry_callbacks_exn(&t_idx);
+
+  for (alloc_idx = 0; alloc_idx < nallocs; alloc_idx++) {
+    unsigned alloc_wosz = encoded_alloc_lens == NULL ? wosize :
+      Wosize_encoded_alloc_len(encoded_alloc_lens[nallocs - 1 - alloc_idx]);
+    unsigned n_samples = 0;
+    alloc_hp -= Whsize_wosize(alloc_wosz);
+    while (alloc_hp < trigger_hp) {
+      n_samples++;
+      trigger_hp -= mt_generate_geom();
+    }
+    if (n_samples > 0) {
+      struct sampled_young_alloc* s;
+      if (allocs_sampled == 1) {
+        /* Found a second sampled allocation! Allocate a buffer for them */
+        sampled_allocs =
+          caml_stat_alloc_noexc(sizeof(*sampled_allocs) * nallocs);
+        sampled_allocs[0] = first_sampled_alloc;
+        res = cb_res;
+        cb_res = caml_alloc(nallocs, 0);
+        Store_field(cb_res, 0, res);
+      }
+      s = &sampled_allocs[allocs_sampled];
+      s->header_ofs = alloc_hp - young_hp;
+      s->wosize = alloc_wosz;
+      s->n_samples = n_samples;
+      callstack = capture_callstack();
```

Comment:
This call can be factored out of the loop.

Reply:
This wouldn't make any difference - the loop almost never runs twice.

Reply:
This will have to be discussed when we will have the actual code for capturing callstacks.

Note, however, that your assertion is wrong if the sampling rate is high. Not that I want to optimize this case, but if that's easy, then why not?